microceph/ceph/rbd_mirror.go

```diff
@@ -0,0 +1,607 @@
+
+	output, err := processExec.RunCommand("rbd", args...)
+	if err != nil {
+		logger.Warnf("failed info operation on res(%s): %v", resource, err)
+		return RbdReplicationImageStatus{State: StateDisabledReplication}, nil
+	}
+
+	err = json.Unmarshal([]byte(output), &response)
+	if err != nil {
+		ne := fmt.Errorf("cannot unmarshal rbd response: %v", err)
+		logger.Errorf(ne.Error())
+		return RbdReplicationImageStatus{State: StateDisabledReplication}, ne
+	}
+
+	logger.Debugf("REPRBD: Image Status: %v", response)
+
+	// Patch required values
+	response.State = StateEnabledReplication
+	response.IsPrimary = strings.Contains(response.Description, "local image is primary")
+
+	return response, nil
+}
+
+// EnablePoolMirroring enables mirroring for an rbd pool in pool mirroring or image mirroring mode.
+func EnablePoolMirroring(pool string, mode types.RbdResourceType, localName string, remoteName string) error {
+	// Enable pool mirroring on the local cluster.
+	err := configurePoolMirroring(pool, mode, "", "")
+	if err != nil {
+		return err
+	}
+
+	// Enable pool mirroring on the remote cluster.
+	err = configurePoolMirroring(pool, mode, localName, remoteName)
+	if err != nil {
+		return err
+	}
+
+	// bootstrap peer
+	return BootstrapPeer(pool, localName, remoteName)
+}
+
+// DisablePoolMirroring disables mirroring for an rbd pool.
+func DisablePoolMirroring(pool string, peer RbdReplicationPeer, localName string, remoteName string) error {
+	// remove peer permissions
+	err := RemovePeer(pool, localName, remoteName)
+	if err != nil {
+		logger.Error(err.Error())
+		return err
+	}
+
+	// Disable pool mirroring on the local cluster.
+	err = configurePoolMirroring(pool, types.RbdResourceDisabled, "", "")
+	if err != nil {
+		logger.Error(err.Error())
+		return err
+	}
+
+	// Disable pool mirroring on the remote cluster.
+	err = configurePoolMirroring(pool, types.RbdResourceDisabled, localName, remoteName)
+	if err != nil {
+		logger.Error(err.Error())
+		return err
+	}
+
+	return nil
+}
+
+// DisableMirroringAllImagesInPool disables mirroring for all images for a pool enabled in pool mirroring mode.
+func DisableMirroringAllImagesInPool(poolName string) error {
+	poolStatus, err := GetRbdMirrorVerbosePoolStatus(poolName, "", "")
+	if err != nil {
+		err := fmt.Errorf("failed to fetch status for %s pool: %v", poolName, err)
+		logger.Error(err.Error())
+		return err
+	}
+
+	disabledImages := []string{}
+	for _, image := range poolStatus.Images {
+		err := disableRbdImageFeatures(poolName, image.Name, []string{"journaling"})
+		if err != nil {
+			return fmt.Errorf("failed to disable journaling on %s/%s", poolName, image.Name)
+		}
+		disabledImages = append(disabledImages, image.Name)
+	}
+
+	logger.Infof("REPRBD: Disabled %v images in %s pool.", disabledImages, poolName)
+	return nil
+}
+
+// getPeerUUID returns the peer ID for the requested peer name.
+func getPeerUUID(pool string, peerName string, client string, cluster string) string {
```

Comment:
Wonder if it wouldn't be better to explicitly handle errors and return `(string, error)` for this func. Afaict callers getting an empty uuid would usually make them fail as well, and with an explicit error var error handling is clearer.

Reply:
ack, changed